File: backend/app/schemas/websocket.py
from typing import Any, Literal, Optional
from pydantic import BaseModel

WSMessageType = Literal[
    "PING",
    "PONG",
    "EVENT",
    "ERROR",
    "SUBSCRIBE",
    "UNSUBSCRIBE",
]

class WSMessage(BaseModel):
    type: WSMessageType
    payload: Any
    timestamp: Optional[int] = None

File: frontend/src/types/websocket.ts
export type WSMessageType =
  | "PING"
  | "PONG"
  | "AUTH"
  | "ERROR"
  | "EVENT"
  | "SUBSCRIBE"
  | "UNSUBSCRIBE";

export interface WSBaseMessage<T = unknown> {
  type: WSMessageType;
  payload: T;
  timestamp?: number;
}

/* Example payloads */
export interface AuthPayload {
  token: string;
}

export interface EventPayload {
  event: string;
  data: unknown;
}

export interface ErrorPayload {
  message: string;
  code?: number;
}
